torch/csrc/distributed/c10d/ProcessGroupNCCL.cpp

```diff
@@ -309,9 +309,28 @@ void cacheAllocatorDeregisterHook(
   }
 }
 
+#ifdef NCCL_COMM_DUMP
 std::string dump_nccl_trace() {
-  return NCCLTraceBuffer::get()->dump();
+  std::unordered_map<
+      std::string /* ncclUniqueID */,
+      std::unordered_map<std::string, std::string> /* dump from this comm */>
+      ncclDumpMap;
+  // dump_nccl_trace is only called from the default PG (uid_=0), but we want to
+  // dump from all comms so we need to iterate over ncclCommDevIdxMap, which
+  // is static
+  ncclCommDevIdxMapMutex.lock();
```

Comment:
any concern this lock would block the dump process? we try to ensure this function would never hang/always dump

Reply:
Good point, I think a safer way is to grab the lock, and do a quick read of all ncclComms, instead of a possible blocking call of ncclCommDump()

Reply:
i was thinking of the opposite case.  we can't acquire ncclCommDevIdxMapMutex and then we hang, preventing us to do the dump.

if we were unable to acquire the lock for some reason, i'd rather we still do the dump and just skip the nccl info

Reply:
the risk of not acquiring the lock 'for some reason' seems low, it was used at the initialization phase and later usage would be all nonblocking.